include/otf2xx/writer/local.hpp

```diff
@@ -160,33 +160,26 @@ namespace writer
             location_.event_written();
         }
 
-        void write(const otf2::event::metric& data)
+        void write(const otf2::event::metric& metric)
         {
-            otf2::reference<otf2::definition::detail::metric_base>::ref_type metric;
+            otf2::reference<otf2::definition::detail::metric_base>::ref_type ref;
 
-            if (data.metric_instance_)
+            if (metric.has_metric_instance())
             {
-                metric = data.metric_instance_->ref();
+                ref = metric.metric_instance_->ref();
             }
             else
             {
-                metric = data.metric_class_->ref();
+                ref = metric.metric_class_->ref();
             }
 
-            std::size_t num_members = data.values().size();
+            std::size_t num_members = metric.values().size();
+            const auto& type_ids = metric.values().type_ids();
+            const auto& metric_values = metric.values().values();
 
-            type_ids_.resize(num_members);
-            values_.resize(num_members);
-
-            for (std::size_t i = 0; i < num_members; i++)
-            {
-                type_ids_[i] = static_cast<OTF2_Type>(data.values()[i].metric->value_type());
-                values_[i] = data.values()[i].value;
-            }
-
-            check(OTF2_EvtWriter_Metric(evt_wrt_, data.attribute_list().get(),
-                                        convert(data.timestamp()), metric, num_members,
-                                        type_ids_.data(), values_.data()),
+            check(OTF2_EvtWriter_Metric(evt_wrt_, metric.attribute_list().get(),
```

Comment:
When I think about this, I'm not convinced. The issue is that this code is indeed faster than the old one. However, it's not optimal in the only writing use case like lo2s.

The metric_values owns the memory which is owned by the event itself. Therefore you're forced to allocate new memory for each event. This is good enough for the rewrite like in HAEC-Sim and the trace merger, but not in lo2s. We need to talk about this in the next meeting.

Reply:
I've been testing the commits from this branch with `lo2s` and have already collected some ideas how remedy those issue. The basic idea is to pre-initialize a `otf2::event::metric` and update it when it needed, then writing it to the archive. This would remove a lot of the `TODO: avoid copy/shared pointers` comments while still using the new code.

Still, the interface of `otf2::event::metric::metric_values` needs some improvement. Especially construction with the correct type information from a `otf2::definition::metric_class` seems necessary.

Reply:
I've discussed this with @tilsche. We concluded that making the event mutable is sufficient.
So add a setter for time in event::base and expose the writable proxy from the metric event such that you can iterate it and directly change the value.